src/xma/src/xmaapi/xmaencoder.cpp

```diff
@@ -308,19 +245,53 @@ xma_enc_session_create(XmaEncoderProperties *enc_props)
     // Create encoder file if it does not exist and initialize all fields 
     xma_enc_session_statsfile_init(enc_session);
 
+    //Obtain lock only for a) singleton changes & b) kernel_info changes
+    std::unique_lock<std::mutex> guard1(g_xma_singleton->m_mutex);
+    //Singleton lock acquired
+
+    if (!kernel_info->soft_kernel && !kernel_info->in_use && !kernel_info->context_opened) {
+        if (xclOpenContext(dev_handle, dev_tmp1.uuid, kernel_info->cu_index_ert, true) != 0) {
+            xma_logmsg(XMA_ERROR_LOG, XMA_ENCODER_MOD, "Failed to open context to CU %s for this session\n", kernel_info->name);
+            free(enc_session->base.plugin_data);
+            free(enc_session);
```

Comment:
While not needed, I always like to set pointers to the memory that is freed to a nullptr.  Makes debugging later in the future easier to do.